### PR TITLE
Restore `SetBreakpoints` API

### PR DESCRIPTION
**Pull request by rjmholt**
https://github.com/PowerShell/PowerShell/pull/11312 changed the `SetBreakpoint()` API in a breaking way. This PR restores the original API while continuing to make the new one available.

Tooling will be impacted by not crashing when targeting the `SetBreakpoints` API across PowerShell versions.

**NOTE** #11312 adds default parameters to other APIs (not broken since they were only added in the 7.0 timeframe). For compatibility it would be favourable to turn those into overloads instead.

Services https://github.com/PowerShell/PowerShell/issues/11619.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tool ing will be impacted.
    - **OR**
    - [x] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

**Comment by TylerLeonhardt** [comment]
I spoke to Dongbo about this back when I implemented #11312 and we decided that that breakage was fine.

**Comment by rjmholt** [comment]
> we decided that that breakage was fine.

Well it's not mutually exclusive -- perhaps we can have both? Also, it feels like such a decision should be documented in either the PR or the extended commit message.

**Comment by TylerLeonhardt** [comment]
Yeah I thought that was captured but I guess not. I'll defer to @daxian-dbw for the right call here. I'm fine either way. Happy to add a comment on my old PR calling out the break.

**Comment by daxian-dbw** [comment]
I agree to have overloads.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/Set-PSBreakpoint.cs
Comment (PoshChan) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/Set-PSBreakpoint.cs]: @rjmholt, your last commit had 7 failures in `PowerShell-CI-windows`
(These are 5 of the failures)

Breakpoint SDK Unit Tests.Managing breakpoints in the host runspace via the SDK.Can set command breakpoints
```powershell
Cannot find an overload for "SetCommandBreakpoint" and the argument count: "1".
at <ScriptBlock>, D:\a\1\s\test\powershell\SDK\Breakpoint.Tests.ps1: line 44
44:             $Host.Runspace.Debugger.SetCommandBreakpoint('Test-ThisCommandDoesNotExist') | Should -BeOfType System.Management.Automation.CommandBreakpoint

```
Breakpoint SDK Unit Tests.Managing breakpoints in the host runspace via the SDK.Can set variable breakpoints
```powershell
Cannot find an overload for "SetVariableBreakpoint" and the argument count: "3".
at <ScriptBlock>, D:\a\1\s\test\powershell\SDK\Breakpoint.Tests.ps1: line 48
48:             $Host.Runspace.Debugger.SetVariableBreakpoint('DebugPreference', 'ReadWrite', { continue }) | Should -BeOfType System.Management.Automation.VariableBreakpoint

```
Breakpoint SDK Unit Tests.Managing breakpoints in the host runspace via the SDK.Can get breakpoints
```powershell
Expected a collection with size 3, but got collection with size 1 @(Line breakpoint on 'D:\a\1\s\test\powershell\SDK\Breakpoint.Tests.ps1:1, 1').
at <ScriptBlock>, D:\a\1\s\test\powershell\SDK\Breakpoint.Tests.ps1: line 56
56:             $Host.Runspace.Debugger.GetBreakpoints() | Should -HaveCount 3
```
Breakpoint SDK Unit Tests.Managing breakpoints in a remote runspace via the SDK.Can set command breakpoints
```powershell
Cannot find an overload for "SetCommandBreakpoint" and the argument count: "2".
at <ScriptBlock>, D:\a\1\s\test\powershell\SDK\Breakpoint.Tests.ps1: line 104
104:             $jobRunspace.Debugger.SetCommandBreakpoint('Write-Verbose', { break }) | Should -BeOfType System.Management.Automation.CommandBreakpoint

```
Breakpoint SDK Unit Tests.Managing breakpoints in a remote runspace via the SDK.Can set variable breakpoints
```powershell
Cannot find an overload for "SetVariableBreakpoint" and the argument count: "3".
at <ScriptBlock>, D:\a\1\s\test\powershell\SDK\Breakpoint.Tests.ps1: line 108
108:             $jobRunspace.Debugger.SetVariableBreakpoint('DebugPreference', 'ReadWrite', { break }) | Should -BeOfType System.Management.Automation.VariableBreakpoint

```

Comment (PoshChan) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/Set-PSBreakpoint.cs]: @rjmholt, your last commit had 7 failures in `PowerShell-CI-macos`
(These are 5 of the failures)

Breakpoint SDK Unit Tests.Managing breakpoints in the host runspace via the SDK.Can set command breakpoints
```powershell
Cannot find an overload for "SetCommandBreakpoint" and the argument count: "1".
at <ScriptBlock>, /Users/runner/runners/2.164.6/work/1/s/test/powershell/SDK/Breakpoint.Tests.ps1: line 44
44:             $Host.Runspace.Debugger.SetCommandBreakpoint('Test-ThisCommandDoesNotExist') | Should -BeOfType System.Management.Automation.CommandBreakpoint

```
Breakpoint SDK Unit Tests.Managing breakpoints in the host runspace via the SDK.Can set variable breakpoints
```powershell
Cannot find an overload for "SetVariableBreakpoint" and the argument count: "3".
at <ScriptBlock>, /Users/runner/runners/2.164.6/work/1/s/test/powershell/SDK/Breakpoint.Tests.ps1: line 48
48:             $Host.Runspace.Debugger.SetVariableBreakpoint('DebugPreference', 'ReadWrite', { continue }) | Should -BeOfType System.Management.Automation.VariableBreakpoint

```
Breakpoint SDK Unit Tests.Managing breakpoints in the host runspace via the SDK.Can get breakpoints
```powershell
Expected a collection with size 3, but got collection with size 1 @(Line breakpoint on '/Users/runner/runners/2.164.6/work/1/s/test/powershell/SDK/Breakpoint.Tests.ps1:1, 1').
at <ScriptBlock>, /Users/runner/runners/2.164.6/work/1/s/test/powershell/SDK/Breakpoint.Tests.ps1: line 56
56:             $Host.Runspace.Debugger.GetBreakpoints() | Should -HaveCount 3
```
Breakpoint SDK Unit Tests.Managing breakpoints in a remote runspace via the SDK.Can set command breakpoints
```powershell
Cannot find an overload for "SetCommandBreakpoint" and the argument count: "2".
at <ScriptBlock>, /Users/runner/runners/2.164.6/work/1/s/test/powershell/SDK/Breakpoint.Tests.ps1: line 104
104:             $jobRunspace.Debugger.SetCommandBreakpoint('Write-Verbose', { break }) | Should -BeOfType System.Management.Automation.CommandBreakpoint

```
Breakpoint SDK Unit Tests.Managing breakpoints in a remote runspace via the SDK.Can set variable breakpoints
```powershell
Cannot find an overload for "SetVariableBreakpoint" and the argument count: "3".
at <ScriptBlock>, /Users/runner/runners/2.164.6/work/1/s/test/powershell/SDK/Breakpoint.Tests.ps1: line 108
108:             $jobRunspace.Debugger.SetVariableBreakpoint('DebugPreference', 'ReadWrite', { break }) | Should -BeOfType System.Management.Automation.VariableBreakpoint

```

Comment (PoshChan) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/Set-PSBreakpoint.cs]: @rjmholt, your last commit had 7 failures in `PowerShell-CI-linux`
(These are 5 of the failures)

Breakpoint SDK Unit Tests.Managing breakpoints in the host runspace via the SDK.Can set command breakpoints
```powershell
Cannot find an overload for "SetCommandBreakpoint" and the argument count: "1".
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/SDK/Breakpoint.Tests.ps1: line 44
44:             $Host.Runspace.Debugger.SetCommandBreakpoint('Test-ThisCommandDoesNotExist') | Should -BeOfType System.Management.Automation.CommandBreakpoint

```
Breakpoint SDK Unit Tests.Managing breakpoints in the host runspace via the SDK.Can set variable breakpoints
```powershell
Cannot find an overload for "SetVariableBreakpoint" and the argument count: "3".
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/SDK/Breakpoint.Tests.ps1: line 48
48:             $Host.Runspace.Debugger.SetVariableBreakpoint('DebugPreference', 'ReadWrite', { continue }) | Should -BeOfType System.Management.Automation.VariableBreakpoint

```
Breakpoint SDK Unit Tests.Managing breakpoints in the host runspace via the SDK.Can get breakpoints
```powershell
Expected a collection with size 3, but got collection with size 1 @(Line breakpoint on '/home/vsts/work/1/s/test/powershell/SDK/Breakpoint.Tests.ps1:1, 1').
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/SDK/Breakpoint.Tests.ps1: line 56
56:             $Host.Runspace.Debugger.GetBreakpoints() | Should -HaveCount 3
```
Breakpoint SDK Unit Tests.Managing breakpoints in a remote runspace via the SDK.Can set command breakpoints
```powershell
Cannot find an overload for "SetCommandBreakpoint" and the argument count: "2".
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/SDK/Breakpoint.Tests.ps1: line 104
104:             $jobRunspace.Debugger.SetCommandBreakpoint('Write-Verbose', { break }) | Should -BeOfType System.Management.Automation.CommandBreakpoint

```
Breakpoint SDK Unit Tests.Managing breakpoints in a remote runspace via the SDK.Can set variable breakpoints
```powershell
Cannot find an overload for "SetVariableBreakpoint" and the argument count: "3".
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/SDK/Breakpoint.Tests.ps1: line 108
108:             $jobRunspace.Debugger.SetVariableBreakpoint('DebugPreference', 'ReadWrite', { break }) | Should -BeOfType System.Management.Automation.VariableBreakpoint

```

**Comment by TravisEz13** [comment]
Please review @PaulHigin

File: src/System.Management.Automation/engine/remoting/server/ServerRunspacePoolDriver.cs
Comment (TylerLeonhardt) [inline src/System.Management.Automation/engine/remoting/server/ServerRunspacePoolDriver.cs]: Shouldn't there be overloads for this? It looks like you just removed the default parameters but this PR is suppose to be a non-functional change.

Comment (rjmholt) [reply src/System.Management.Automation/engine/remoting/server/ServerRunspacePoolDriver.cs]: I discussed it with @daxian-dbw and the breakdown is:

- Providing overloads for each default parameter would mean 6 or 24 overloads in some cases
- Although default parameters make sense for a PowerShell cmdlet, this is an API, so the UX is different; pith is not as valuable and may not actually be useful anyway
- In particular I noticed that
  - in the Set-PSBreakpoint case, the defaults were only used once and in that case it seemed to me that setting `path: null` actually seemed better
  - in PSES, we use reflection to call the methods, so we don't benefit at all from the default parameters
  - the inheriting classes also needed to set default parameters. Now they just implement the main virtual method
- Given the unusualness of needing to target a different runspace, and the way I'd already overloaded the other APIs, I kept the `runspaceId` overloads
- Each parameter is documented (I updated the documentation) so that API consumers will know if they should set parameters as `null`

Comment (TylerLeonhardt) [reply src/System.Management.Automation/engine/remoting/server/ServerRunspacePoolDriver.cs]: I'm fine with this but it seems like you're now breaking the API (which I understand I already broke) for the sake of consistency?

That's not typically the approach we've taken to breaking changes.

> Providing overloads for each default parameter would mean 6 or 24 overloads in some cases

This is one of the reasons why the addition of the `runspaceId` with a default parameter was done and since there were already default parameters, it fit right in - despite it being a breaking change.

Comment (rjmholt) [reply src/System.Management.Automation/engine/remoting/server/ServerRunspacePoolDriver.cs]: > but it seems like you're now breaking the API

Yeah that's true in the sense that I'm changing the API, but it was only introduced in a 7 preview; from the perspective of previously GA's PowerShells (6.2, 5.1) this PR restores one API and doesn't break any others.

> for the sake of consistency?

An inconsistent API would be ok, but the default parameters [can lead to runtime exceptions in consuming code if we ever add more APIs](https://haacked.com/archive/2010/08/10/versioning-issues-with-optional-arguments.aspx/). If we ever decided we needed to provide more overloads or parameters for these breakpoint APIs, we'd need to tread very carefully if we keep the default parameters.

I don't really like the extent of these changes, but they try to set API to something we can maintain and extend before 7 is locked in place.

Comment (TylerLeonhardt) [reply src/System.Management.Automation/engine/remoting/server/ServerRunspacePoolDriver.cs]: > this PR restores one API and doesn't break any others

you're making parameters that use to have a default value to not having a default value. That's breaking the previous GA API, no?

Comment (rjmholt) [reply src/System.Management.Automation/engine/remoting/server/ServerRunspacePoolDriver.cs]: > you're making parameters that use to have a default value to not having a default value

Yes, absolutely.

> That's breaking the previous GA API, no?

Not quite, because that API was introduced in https://github.com/PowerShell/PowerShell/commit/cc0fed479a3e455b746a3d12597f078462f2d644, released in 7.0.0-preview.4. So the API's never GA'd.

This change is like the `-loadprofile` -> `-login` change, where it was released in a preview version but never GA'd.

Comment (daxian-dbw) [reply src/System.Management.Automation/engine/remoting/server/ServerRunspacePoolDriver.cs]: This PR is marked with `GA-Consider` milestone, and once approved will be included in GA release.

Comment (TylerLeonhardt) [reply src/System.Management.Automation/engine/remoting/server/ServerRunspacePoolDriver.cs]: Ah so you've only added the overloads to the previously GA'd APIs. That was the detail I was missing.

I'm ok with this then.

**Review by PaulHigin — APPROVED** [review]

File: src/System.Management.Automation/engine/remoting/server/ServerRunspacePoolDriver.cs
Comment (TravisEz13) [reply src/System.Management.Automation/engine/remoting/server/ServerRunspacePoolDriver.cs]: @TylerLeonhardt is this resolved?

Comment (TylerLeonhardt) [reply src/System.Management.Automation/engine/remoting/server/ServerRunspacePoolDriver.cs]: Yes it is

**Review by TravisEz13 — APPROVED** [review]
Approving as maintainer

**Pull request merged**
